### PR TITLE
ci: ROCm gfx1151 fixes (draft, stacked on #26461)

### PR DESCRIPTION
**Pull request by jimw567**
## Overview

This branch is based on #26461's head, so the diff currently includes that PR's commits too. It adds three changes needed to get the `gpu-rocm` (gfx1151 / Strix Halo) self-hosted job green:

1. **CUDA: allow integrated-GPU host output buffer in debug assert** — on integrated RDNA3.5 the scheduler can place a node's output on the host-visible buffer; the debug assert in `ggml_cuda_graph_evaluate_and_capture` rejected it, aborting `ctest_debug`. Extends the existing `integrated && cuda_host` exception (already applied to src tensors) to the node's own output buffer. Debug-only.

2. **ci: enable unified memory for the ROCm gfx1151 job** — on gfx1151, GPU kernels reading mmap-loaded weights can return incorrect output (intermittent, ~0.3 NMSE vs CPU) due to a host/device coherence issue; `--no-mmap`, `HIP_LAUNCH_BLOCKING=1`, and `GGML_CUDA_ENABLE_UNIFIED_MEMORY=1` all avoid it. The CI job sets `GGML_CUDA_ENABLE_UNIFIED_MEMORY=1` (managed memory), which keeps mmap and restores coherence. Workaround pending a proper ROCm/HIP fix.

3. **test-llama-archs: skip jamba on HIP** — jamba produces ~0.55 NMSE vs CPU on gfx1151 (a separate SSM-kernel accuracy bug, unaffected by the coherence workaround). Skipped via `#ifdef GGML_USE_HIP` in `arch_supported()`, matching the existing WebGPU per-arch carve-out, so the test still runs for the other ~40 architectures.

## Requirements

- I have read and agree with the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md)
- AI usage disclosure: Yes. AI helped me understand the issues and iterate different ways to fix the issue. All ideas are directed by me and solutions are reviewed and approved by me.

### CODE REVIEW

**Comment by jimw567** [comment]
@taronaeo This PR needs approval for the workflow to start. Please help approve it.

**Comment by taronaeo** [comment]
I have just approved it :)

**Comment by taronaeo** [comment]
@jimw567 Looks like the CI is still failing, but now for a different reason. For some reason it is exiting the runner with

```
Process completed with exit code 20.
```

And I don't see any errors from the logs as well.

image

**Comment by jimw567** [comment]
> @jimw567 Looks like the CI is still failing, but now for a different reason. For some reason it is exiting the runner with
> 
> ```
> Process completed with exit code 20.
> ```
> 
> And I don't see any errors from the logs as well.
> 
> image

As a regular user I can't see any log. Is that disabled on purpose? In any case I am debugging it from the runner.

image

**Comment by taronaeo** [comment]
I can't see the logs after the job has failed too. I had to stay on the page while the runner was running to watch the logs fail. I think this is something new that GitHub did because we used to be able to see the failure.

**Comment by jimw567** [comment]
@taronaeo I pushed two commits to this PR. The full CI has been validated locally with these two commits. The goal of this PR is to enable he gpu-rocm runner for the CI, so I worked around some CI issues. I will raise separate PRs to remove those workarounds as the work planned has nothing to do with enabling the runner. Key changes and workarounds:

- [ ]  use HIP_LAUNCH_BLOCKING for ROCm gfx1151 job to fix wrong inference output on gfx1151 (qwen3 perplexity ~88 -> 9.44). 
- [ ] skip top-k subtests on HIP backend  test tests/test-backend-sampler.cpp: under as ROCm has no CUB, so TOP_K/ARGSORT is unsupported at vocab scale (bitonic argsort capped at ncols<=1024)

The new push doesn't seem to trigger a rerun of CI. Can you please approve it again? Having said that, would it be possible to whitelist my account so that my upstream PRs can trigger CI runs automatically?

**Comment by jimw567** [comment]
> I can't see the logs after the job has failed too. I had to stay on the page while the runner was running to watch the logs fail. I think this is something new that GitHub did because we used to be able to see the failure.

That is a bummer. Luckily I have access to the runner to check all the logs.

**Comment by taronaeo** [comment]
Its green now 🎉

**Comment by jimw567** [comment]
> Its green now 🎉

Yeah!!! I will raise a PR to your original PR so we get this runner properly added.

**Comment by jimw567** [comment]
@taronaeo Looks like I can't raise a PR against your fork.  If it's ok with you, we can merge this PR 26544 as it is. It includes all changes from your branch:   taronaeo:ci/amd-rocm-hip

image

**Comment by taronaeo** [comment]
> @taronaeo Looks like I can't raise a PR against your fork.  If it's ok with you, we can merge this PR 26544 as it is. It includes all changes from your branch:   taronaeo:ci/amd-rocm-hip
> 
> 
> 
> image
> 
> 

image

I am able to do it but it appears our baseline is different, and hence can't automatically merge.

I am OK with using this PR as the correct one now. Whenever you are ready, mark the PR as ready and I'll ping the maintainers to review this one instead :)

**Comment by jimw567** [comment]
@taronaeo Thanks! I just marked the PR for reviewing.

File: tests/test-backend-sampler.cpp [hunk 1]
```diff
@@ -1668,9 +1668,23 @@ static std::vector<const backend_test_case *> collect_tests_to_run(const std::st
...
+            }
+#ifdef GGML_USE_HIP
+            // The ROCm backend does not support the TOP_K / ARGSORT op at vocab
+            // scale (no CUB; bitonic argsort is capped at ncols <= 1024), so
+            // top-k / top-p backend samplers cannot be offloaded. These tests
+            // assert that the offload happened (exact top-k logit count, filtered
+            // top-p logits, mid-stream top-k reconfigure, top-k + penalties), so
+            // they fail on HIP. Skip until TOP_K is supported on the ROCm backend.
+            if (test.name == "penalties"   || test.name == "set_sampler" ||
+                test.name == "mixed"       || test.name == "top_p") {
```
Comment (taronaeo) [inline tests/test-backend-sampler.cpp]:
```suggestion
            if (test.name == "penalties" || test.name == "set_sampler" ||
                test.name == "mixed"     || test.name == "top_p") {
```

nit

**Review by taronaeo — APPROVED** [review]

**Comment by taronaeo** [comment]
Btw is this failing CI of any concern? 

https://github.com/ggml-org/llama.cpp/actions/runs/30984084124/job/92238741006#step:7:1

cc: @CISC for another review

File: tests/test-backend-sampler.cpp [hunk 2]
```diff
@@ -1668,9 +1668,23 @@ static std::vector<const backend_test_case *> collect_tests_to_run(const std::st
...
+            if (!test.enabled_by_default) {
+                continue;
+            }
+#ifdef GGML_USE_HIP
+            // The ROCm backend does not support the TOP_K / ARGSORT op at vocab
+            // scale (no CUB; bitonic argsort is capped at ncols <= 1024), so
+            // top-k / top-p backend samplers cannot be offloaded. These tests
+            // assert that the offload happened (exact top-k logit count, filtered
+            // top-p logits, mid-stream top-k reconfigure, top-k + penalties), so
+            // they fail on HIP. Skip until TOP_K is supported on the ROCm backend.
```
Comment (ggerganov) [inline tests/test-backend-sampler.cpp]:
```suggestion
            // TODO: remove this when https://github.com/ggml-org/llama.cpp/pull/26592 is merged
```

File: .github/workflows/build-self-hosted.yml [hunk 3]
```diff
@@ -71,6 +71,26 @@ jobs:
...
+      - name: Clone
+        id: checkout
+        uses: actions/checkout@v6
+
+      - name: Test
+        id: ggml-ci
+        # HIP_LAUNCH_BLOCKING=1: workaround for an async-execution correctness
+        # issue on integrated RDNA3.5 (gfx1151) where batched inference returns
+        # incorrect output (perplexity ~88 vs ~9.4). Serializing kernel launches
+        # restores correctness. Remove once the underlying ROCm/HIP issue is fixed.
```
Comment (ggerganov) [inline .github/workflows/build-self-hosted.yml:87]: Is the ROCm/HIP issue tracked somewhere?

Comment (jimw567) [reply .github/workflows/build-self-hosted.yml:87]: Yes, it's being tracked in our internal JIRA system (see the snapshot below). 

image

File: tests/test-backend-sampler.cpp (hunk 2 above)
Comment (jimw567) [reply tests/test-backend-sampler.cpp]: Thanks. Didn't realize there is already a PR in the pipeline to fix the top_k issue.

**Comment by CISC** [comment]
> Btw is this failing CI of any concern?
> 
> https://github.com/ggml-org/llama.cpp/actions/runs/30984084124/job/92238741006#step:7:1

@jimw567 Please check.

**Comment by IMbackK** [comment]
> Btw is this failing CI of any concern?
> 
> https://github.com/ggml-org/llama.cpp/actions/runs/30984084124/job/92238741006#step:7:1
> 
> cc: @CISC for another review

no its a known failure, need to add this kernel to the ignore list (and see if we can avoid the spill at a later time)

**Review by CISC — APPROVED** [review]

**Comment by CISC** [comment]
> I can't see the logs after the job has failed too. I had to stay on the page while the runner was running to watch the logs fail. I think this is something new that GitHub did because we used to be able to see the failure.

This is frustrating, its value as a CI is not that great if we can't view the logs, please look into this.

**Comment by jimw567** [comment]
> > I can't see the logs after the job has failed too. I had to stay on the page while the runner was running to watch the logs fail. I think this is something new that GitHub did because we used to be able to see the failure.
> 
> This is frustrating, its value as a CI is not that great if we can't view the logs, please look into this.

Totally agree . I actually looked into this two days ago. This looks like an issue with our firewall preventing logs uploade.  I am working with our IT to update firewall policy to address the issue. Having said that I did put some temp solution in the PR to keep the log around on the runner for each run. I will be more than happy to provide the log to help debug if needed.

**Comment by ggerganov** [comment]
> > > I can't see the logs after the job has failed too. I had to stay on the page while the runner was running to watch the logs fail. I think this is something new that GitHub did because we used to be able to see the failure.
> > 
> > 
> > This is frustrating, its value as a CI is not that great if we can't view the logs, please look into this.
> 
> Totally agree . I actually looked into this two days ago. This looks like an issue with our firewall preventing logs uploade. I am working with our IT to update firewall policy to address the issue. Having said that I did put some temp solution in the PR to keep the log around on the runner for each run. I will be more than happy to provide the log to help debug if needed.

@jimw567 Any progress on the firewall issue?

**Comment by jimw567** [comment]
> > > > I can't see the logs after the job has failed too. I had to stay on the page while the runner was running to watch the logs fail. I think this is something new that GitHub did because we used to be able to see the failure.
> > > 
> > > 
> > > This is frustrating, its value as a CI is not that great if we can't view the logs, please look into this.
> > 
> > 
> > Totally agree . I actually looked into this two days ago. This looks like an issue with our firewall preventing logs uploade. I am working with our IT to update firewall policy to address the issue. Having said that I did put some temp solution in the PR to keep the log around on the runner for each run. I will be more than happy to provide the log to help debug if needed.
> 
> @jimw567 Any progress on the firewall issue?

@ggerganov I just followed up with our IT and will post an update when I hear something.

**Comment by jimw567** [comment]
@ggerganov and @taronaeo Just so you know that our DevOps has enabled log upload to github. I just checked it's working as expected now.

https://github.com/ggml-org/llama.cpp/actions/runs/33734279996/job/100682418244?pr=28279

With this, can we make rocm checks gating in that a PR can't be merged if rocm check fails?

**Comment by taronaeo** [comment]
Great! Thanks for letting us know.

> With this, can we make rocm checks gating in that a PR can't be merged if rocm check fails?

I don't know if GitHub has such a feature, probably has, but that would also affect all other falsely-failing CIs and make the PR un-mergable. In general we usually check why the CI failed in the first place before choosing to merge, and if we still choose to merge anyway, we usually provide a reason.

**Comment by jimw567** [comment]
> Great! Thanks for letting us know.
> 
> > With this, can we make rocm checks gating in that a PR can't be merged if rocm check fails?
> 
> I don't know if GitHub has such a feature, probably has, but that would also affect all other falsely-failing CIs and make the PR un-mergable. In general we usually check why the CI failed in the first place before choosing to merge, and if we still choose to merge anyway, we usually provide a reason.

ok. Thanks for the explanation. Understood. Is is possible to mention @ggml-org/amd when rocm CI fails and you decide to merge it? This will give us a chance to respond.

**Comment by superm1** [comment]
> I don't know if GitHub has such a feature, probably has

There is a concept of `required` jobs that you can't merge if they're red.  It's of course a double edged sword.

> but that would also affect all other falsely-failing CIs and make the PR un-mergable

Overall if a job fails every time and is expected to fail IMO it's better to disable it and then only re-enable it when you know it will be reliable.  Otherwise, it's just noise that everyone ignores and a waste of resources.

> Is is possible to mention @ggml-org/amd when rocm CI fails and you decide to merge it? This will give us a chance to respond.

Yes, please.  This would be really helpful.
